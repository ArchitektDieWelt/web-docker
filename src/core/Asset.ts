export type AssetPosition = "head" | "body";

export type AssetBase = {
  src: string;
  name?: string;
  position?: AssetPosition;
  priority?: number;
  countries?: string[];
};

export type AssetScript = AssetBase & {
  type: "js";
  async?: boolean;
  buildType?: "modern" | "legacy";
  deferred?: boolean;
};

export type AssetLink = AssetBase & {
  type: "css";
  media?: string;
};

export type Asset = AssetLink | AssetScript;
