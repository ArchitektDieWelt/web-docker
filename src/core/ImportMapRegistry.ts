interface ImportMap {
  [key: string]: string;
}

export class ImportMapRegistry {
  private importMap: ImportMap = {};

  constructor(
    private readonly script: HTMLScriptElement = document.createElement("script"),
    private readonly documentHead = document.head
  ) {
    this.script.type = "importmap";
    this.documentHead.appendChild(this.script);
  }

  public add(importMap: ImportMap) {
    this.importMap = {
      ...this.importMap,
      ...importMap,
    };
    this.script.textContent = JSON.stringify(this.importMap);
  }

  public get(): ImportMap {
    return this.importMap;
  }
}
