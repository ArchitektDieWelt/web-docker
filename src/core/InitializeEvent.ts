
export const InitializeEventType = "web-docker:initialize" as const;

class InitializeEvent extends CustomEvent<{
  configFilePath: string;
  logEvents: boolean;
}> {
  constructor(detail: { configFilePath: string; logEvents: boolean }) {
    super(InitializeEventType, { detail });
  }
}

export default InitializeEvent;
