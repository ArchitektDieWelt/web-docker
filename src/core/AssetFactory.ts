import { Asset, AssetLink, AssetPosition, AssetScript } from "~/core/Asset";

export default class AssetFactory {
  private static buildTagScript(attr: AssetScript): HTMLScriptElement {
    const script = document.createElement("script");
    script.src = attr.src;
    if (attr.buildType === "modern") {
      script.type = "module";
      script.setAttribute("crossorigin", "");
    } else {
      script.type = "text/javascript";
      if (attr.async) script.setAttribute("async", "");
    }

    if (attr.priority)
      script.setAttribute("data-priority", attr.priority.toString());

    return script;
  }

  private static buildTagLink(attr: AssetLink): HTMLLinkElement {
    const link = document.createElement("link");
    link.setAttribute("rel", "stylesheet");
    link.setAttribute("href", attr.src);
    if (attr.media) link.setAttribute("media", attr.media);

    if (attr.priority)
      link.setAttribute("data-priority", attr.priority.toString());

    return link;
  }

  public create(
    assets: Asset[],
    position: AssetPosition = "head",
    countryCode?: string
  ): (HTMLLinkElement | HTMLScriptElement)[] {
    return (
      countryCode
        ? assets.filter(
            (asset) =>
              typeof asset.countries !== "object" ||
              asset.countries.includes(countryCode)
          )
        : assets
    )
      .filter((asset) => (asset.position ?? "head") === position)
      .sort((a, b) => {
        if (!a.priority || !b.priority) return 0;
        if (a.priority < b.priority) {
          return 1;
        } else if (a.priority > b.priority) {
          return -1;
        }
        return 0;
      })
      .map((entry) => {
        switch (entry.type) {
          case "css":
            return AssetFactory.buildTagLink(entry);
          case "js":
            return AssetFactory.buildTagScript(entry);
        }
      })
      .filter((item): item is HTMLLinkElement | HTMLScriptElement => !!item);
  }
}
