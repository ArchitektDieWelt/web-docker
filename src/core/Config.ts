import { ModuleConfig } from "~/core/ModuleConfig";

export type Config = ModuleConfig;
