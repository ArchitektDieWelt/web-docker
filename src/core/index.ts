import RegisterEvent, { RegisterEventType } from "~/core/RegisterEvent";
import InitializeEvent, { InitializeEventType } from "~/core/InitializeEvent";

export { default as initialize } from "./initialize";

declare global {
  interface WindowEventMap {
    [RegisterEventType]: RegisterEvent;
    [InitializeEventType]: InitializeEvent;
  }
  interface Window {
    openreply: {
      initialize: (config: {
        configFilePath: string;
      }) => void;
    };
  }
}
