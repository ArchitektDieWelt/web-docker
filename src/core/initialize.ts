import {
  RemoteConfig,
  RemoteConfigurationService,
} from "~/core/RemoteConfigurationService";
import ModuleRegistry from "~/core/ModuleRegistry";
import { Config } from "~/core/Config";
import { ModuleConfigService } from "~/core/ModuleConfig";
import { RegisterEventType } from "~/core/RegisterEvent";
import { InitializeEventType } from "~/core/InitializeEvent";

const initialize = async (remoteConfig: RemoteConfig) => {
  const moduleConfigService = new ModuleConfigService();
  const registry = new ModuleRegistry();
  const remoteConfigurationService = new RemoteConfigurationService(
    remoteConfig
  );
  // NOTE: event registry has to be done before any async operation. Since Async operations
  // do not block the event loop and the registration event might be dispatched while another event loop is running.
  window.addEventListener(
    RegisterEventType,
    (event: CustomEvent<Config>) => {
      const config = moduleConfigService.getModuleConfig(event.detail);
      registry.addReplace(config);
    }
  );

  const remoteConfigurations = await remoteConfigurationService.fetch();
  remoteConfigurations?.forEach((config: Config) => {
    const moduleConfig = moduleConfigService.getModuleConfig(config);
    registry.add(moduleConfig);
  });
};

window.openreply = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<RemoteConfig>) => {
    await initialize(event.detail);
  }
);

export default initialize;
