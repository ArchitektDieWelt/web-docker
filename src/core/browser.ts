import { InitializeEventType } from "~/core/InitializeEvent";
import initialize, { WebDockerOptions } from "~/core/initialize";

window.webdocker = { initialize };

window.addEventListener(
  InitializeEventType,
  async (event: CustomEvent<WebDockerOptions>) => {
    await initialize(event.detail);
  }
);

