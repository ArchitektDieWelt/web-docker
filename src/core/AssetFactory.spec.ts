import { describe, it, expect } from "vitest";
import AssetFactory from "~/core/AssetFactory";
import { AssetLink, AssetScript } from "~/core/Asset";

describe("AssestFactory", () => {
  it("constructs script asset", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetScript = {
      src: "src/",
      type: "js",
    };

    const element = assetFactory.create([asset]);

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });

  it("constructs link asset", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetLink = {
      src: "/src/",
      type: "css",
    };

    const element = assetFactory.create([asset]);

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });

  it("constructs multiple assets of different types", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetLink = {
      src: "/src/",
      type: "css",
    };

    const asset2: AssetScript = {
      src: "/src/",
      type: "js",
    };

    const element = assetFactory.create([asset, asset2]);

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });

  it("ignores the asset if it does not match the position argument", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetLink = {
      src: "/src/",
      type: "css",
      position: "head",
    };

    const asset2: AssetScript = {
      src: "/src/",
      type: "js",
      position: "body",
    };

    const element = assetFactory.create([asset, asset2], "head");

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });

  it("assumes default head position for assets if the asset does not specify", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetLink = {
      src: "/src/",
      type: "css",
    };

    const element = assetFactory.create([asset], "head");

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });

  it("creates list in order of asset priority", () => {
    const assetFactory = new AssetFactory();
    const asset: AssetScript = {
      src: "/src/",
      type: "js",
      priority: 1,
    };

    const asset2: AssetLink = {
      src: "/src2/",
      type: "css",
      priority: 2,
    };

    const element = assetFactory.create([asset, asset2], "head");

    expect(element).not.toBeNull();

    expect(element).toMatchSnapshot();
  });
});
