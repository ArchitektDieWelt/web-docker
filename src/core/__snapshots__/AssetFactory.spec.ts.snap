// Vitest Snapshot v1

exports[`AssestFactory > assumes default head position for assets if the asset does not specify 1`] = `
[
  <link
    href="/src/"
    rel="stylesheet"
  />,
]
`;

exports[`AssestFactory > constructs link asset 1`] = `
[
  <link
    href="/src/"
    rel="stylesheet"
  />,
]
`;

exports[`AssestFactory > constructs multiple assets of different types 1`] = `
[
  <link
    href="/src/"
    rel="stylesheet"
  />,
  <script
    src="/src/"
    type="text/javascript"
  />,
]
`;

exports[`AssestFactory > constructs script asset 1`] = `
[
  <script
    src="src/"
    type="text/javascript"
  />,
]
`;

exports[`AssestFactory > creates list in order of asset priority 1`] = `
[
  <link
    data-priority="2"
    href="/src2/"
    rel="stylesheet"
  />,
  <script
    data-priority="1"
    src="/src/"
    type="text/javascript"
  />,
]
`;

exports[`AssestFactory > ignores the asset if it does not match the position argument 1`] = `
[
  <link
    href="/src/"
    rel="stylesheet"
  />,
]
`;
